assessment-api/assessment-actors/src/main/scala/org/sunbird/managers/CopyManager.scala

```diff
@@ -0,0 +1,409 @@
+          put(AssessmentConstants.ROOT, true.asInstanceOf[AnyRef])
+          put(AssessmentConstants.PRIMARY_CATEGORY, node.getMetadata.get(AssessmentConstants.PRIMARY_CATEGORY))
+        }
+      })
+      populateHierarchyRequest(children, nodesModified, hierarchy, node.getIdentifier, copyType, request, idMap)
+      getExternalData(idMap.keySet().asScala.toList, request).map(exData => {
+        idMap.asScala.toMap.foreach(entry => {
+          nodesModified.get(entry._2).asInstanceOf[java.util.Map[String, AnyRef]].get("metadata").asInstanceOf[util.Map[String, AnyRef]].putAll(exData.getOrDefault(entry._1, new util.HashMap[String, AnyRef]()).asInstanceOf[util.Map[String, AnyRef]])
+        })
+        new util.HashMap[String, AnyRef]() {
+          {
+            put(AssessmentConstants.NODES_MODIFIED, nodesModified)
+            put(AssessmentConstants.HIERARCHY, hierarchy)
+          }
+        }
+      })
+
+    } else Future(new util.HashMap[String, AnyRef]())
+  }
+
+  def getExternalData(identifiers: List[String], request: Request)(implicit ec: ExecutionContext, oec: OntologyEngineContext): Future[util.Map[String, AnyRef]] = {
+    val req = new Request(request)
+    req.getContext().putAll(Map("objectType" -> "Question", "schemaName" -> "question").asJava)
+    req.put("identifiers", identifiers)
+    val result = new util.HashMap[String, AnyRef]()
+    val externalProps = DefinitionNode.getExternalProps(req.getContext.getOrDefault("graph_id", "domain").asInstanceOf[String], req.getContext.getOrDefault("version", "1.0").asInstanceOf[String], req.getContext.getOrDefault("schemaName", "question").asInstanceOf[String])
+    val externalPropsResponse = oec.graphService.readExternalProps(req, externalProps)
+    externalPropsResponse.map(response => {
+      identifiers.map(id => {
+        val externalData = Optional.ofNullable(response.get(id).asInstanceOf[util.Map[String, AnyRef]]).orElse(new util.HashMap[String, AnyRef]())
+        result.put(id, externalData)
+      })
+      result
+    })
+  }
+
+  def populateHierarchyRequest(children: util.List[util.Map[String, AnyRef]], nodesModified: util.HashMap[String, AnyRef], hierarchy: util.HashMap[String, AnyRef], parentId: String, copyType: String, request: Request, idMap: java.util.Map[String, String])(implicit ec: ExecutionContext, oec: OntologyEngineContext): Unit = {
+    if (null != children && !children.isEmpty) {
+      children.asScala.toList.foreach(child => {
+        val id = if ("Parent".equalsIgnoreCase(child.get(AssessmentConstants.VISIBILITY).asInstanceOf[String])) {
+          val identifier = UUID.randomUUID().toString
+          nodesModified.put(identifier, new util.HashMap[String, AnyRef]() {
+            {
+              put(AssessmentConstants.METADATA, cleanUpCopiedData(new util.HashMap[String, AnyRef]() {
+                {
+                  putAll(child)
+                  put(AssessmentConstants.COPY_OF, child.getOrDefault(AssessmentConstants.IDENTIFIER,""))
```

Comment:
Is this metadata COPY_OF getting cleaned before making updateHierarchy call?

Reply:
Yes. It is getting cleaned up in generateBranchingRecord function of BranchingUtil